Source/WebCore/editing/MarkupAccumulator.cpp

```diff
@@ -387,9 +388,15 @@ const ShadowRoot* MarkupAccumulator::suitableShadowRoot(const Node& node)
 
 void MarkupAccumulator::startAppendingNode(const Node& node, Namespaces* namespaces)
 {
-    if (RefPtr element = dynamicDowncast<Element>(node))
+    if (RefPtr element = dynamicDowncast<Element>(node)) {
         appendStartTag(m_markup, *element, namespaces);
-    else if (auto* shadowRoot = suitableShadowRoot(node)) {
+
+        // Currently URL Replacement only happens when saving markup to disk and the file uses UTF-8 encoding.
+        // To ensure the file can be loaded correctly, change the specified encoding to UTF-8.
+        if (m_urlReplacementData && is<HTMLHeadElement>(element))
+            m_markup.append("<meta charset=\"UTF-8\"><!-- Encoding specified by WebKit -->"_s);
```

Comment:
How about `<html charset=~>` and `<?xml encoding=~>`?

Reply:
The cases where we encounter the issue so far are all using meta tag to declare character encoding (rdar://122425575) (and this seems the recommended way to declare encoding by spec), so this workaround should resolve most of them and should be good enough for now.
Thanks for the review!

Reply:
I don't think so. We most certainly need to take care of those two cases.

Reply:
As discussed offline, there is no `<html charset=~>` and xml is not a big use case for this feature now. We will polish the heuristics later. I added a FIXME here.